"List my articles" endpoints, serializations to different Article types

andyhaskell opened the pull request:
This PR makes the following additions to the Go client:

* Add methods for "list my articles" endpoints (`/articles/me/*`) for listing the authenticated user's articles
* Add convenience methods to the `/articles` endpoints for listing articles by tag and user so these functionalities can be used without filling out a struct
* Define types for the variations on the `Article` type for the list endpoints, POST/PUT payloads

andyhaskell committed:
Additional error handling for HTTP roundtrips, JSON unmarshalling

andyhaskell committed:
Convenience methods for listing articles by tag and by user

These methods provide more specific sets of query parameters to the
list articles endpoint to make these use cases simpler to work with.

andyhaskell committed:
Add methods for "list my articles" endpoints

These endpoints are called at /api/articles/me/* to retrieve articles
made by the caller.

andyhaskell committed:
Separate types for listed articles, POST request articles

This is so that the different serializations for these endpoints are
reflected.

VictorAvelar commented:
@andyhaskell thanks for your PR, would you mind adding some tests and fixing the existing ones, please?

andyhaskell committed:
Fix test coverage for the list articles endpoint with query params

The new error handling in json.Unmarshal on the List method caused this
test to error, so to fix the test coverage, this commit has the mock
list articles endpoint return a JSON response for the client to
unmarshal,

andyhaskell commented:
Sounds great, just fixed the new tests, so I'll work on some tests for the new endpoints I added

andyhaskell committed:
Test coverage for "list my articles" endpoints

Additionally, add test coverage for convenience methods on /articles
endpoints, ArticleUpdate deserialization on POST and PUT endpoints

VictorAvelar commented:
@andyhaskell is everything ready from your side? I would like to merge your PR.
